src/dune_tui/dune_tui.ml

```diff
@@ -144,35 +33,304 @@ module Tui = struct
+        ; I.uchar helper_attr Unicode.ogham_feather_mark 1 1
+        ]
     in
-    let messages =
-      List.map messages ~f:(fun msg ->
-        image_of_user_message_style_pp (User_message.pp msg))
+    let hsnap_or_leave ~width img =
+      if I.width img < width then I.hsnap ~align:`Middle width img else img
     in
-    Notty.I.vcat (messages @ status)
+    Ui.zcat
+      [ Ui.atom
+        @@ I.zcat [ hsnap_or_leave ~width:w status; horiztonal_rule ~attr:helper_attr ~w ]
+      ; help_button
+      ]
+  ;;
+
+  (** Our document has 3 components:
+      - A help box
+      - A status bar
+      - A message viewer *)
+  let document =
+    let* { ui = help_box; toggle = handle_help } = help_box in
+    let+ status_bar = status_bar
+    and+ message_viewer = Message_viewer.ui in
+    let keyboard_handler = function
+      (* When we encounter q we make sure to quit by signaling termination. *)
+      | `ASCII 'q', _ ->
+        Unix.kill (Unix.getpid ()) Sys.sigterm;
+        `Handled
+      (* Toggle help screen *)
+      | `ASCII '?', _ ->
+        handle_help ();
+        `Handled
+      | _ -> `Unhandled
+    in
+    Ui.zcat [ Ui.vcat [ status_bar; message_viewer ]; help_box ]
+    |> Ui.keyboard_area keyboard_handler
+  ;;
+
+  let renderer = Renderer.make ()
+
+  let update_peristant_state =
+    let update equal v x = if equal (Lwd.peek v) x then () else Lwd.set v x in
+    fun (state : Dune_threaded_console.state) ->
+      let size = Term.size (term ()) in
+      update (Tuple.T2.equal Int.equal Int.equal) term_size size;
+      update
+        (Option.equal (fun x y ->
+           Ordering.is_eq (Pp.compare ~compare:User_message.Style.compare x y)))
+        status_line
+        state.status_line;
+      if let l = Lwd.peek messages in
+         List.length l = Queue.length state.messages
+         && List.equal User_message.equal l (Queue.to_list state.messages)
+      then ()
+      else Lwd.set messages (Queue.to_list state.messages)
   ;;
 
   let render (state : Dune_threaded_console.state) =
-    let messages = Queue.to_list state.messages in
-    let image = image ~status_line:state.status_line ~messages in
+    let size = Term.size (term ()) in
+    (* Update the persistant values tracked by other components. *)
+    update_peristant_state state;
+    (* This is a standard [Lwd] routine for creating a document. *)
+    let root = Lwd.observe document in
+    let image =
+      let rec stabilize () =
+        let tree = Lwd.quick_sample root in
+        Renderer.update renderer size tree;
+        let image = Renderer.image renderer in
+        if Lwd.is_damaged root then stabilize () else image
+      in
+      stabilize ()
+    in
+    (* Finally we use Notty to show the image. *)
     Term.image (term ()) image
   ;;
 
-  let resize mutex (state : Dune_threaded_console.state) =
+  (** Update any global state and finish *)
+  let finish_dirty_interaction ~mutex (state : Dune_threaded_console.state) =
```

Comment:
Call this function `set_dirty` and inline the `Unix.gettimeoday ()` at all the callers.